Add MySQL compression level variable and change default to compression level 3 (v3.0)

yashwantsahu20 opened the pull request:
Default compression level was 6 for MySQL clients  and this was making ProxySQL slower.

Following changes has been made:
1. Made default compression level as 3.
2. Added MySQL compression level variable

For detail of issue, #4721

yashwantsahu20 committed:
Added  MySQL compression level

yashwantsahu20 committed:
1. Changed name to protocol_compression_level.
2. Added TAP test

yashwantsahu20 committed:
Added test for compression level variable

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+	mysql_query(proxysql_admin, query.c_str());
+}
+
+int32_t get_compression_level(MYSQL* proxysql_admin) {
```
We have several utility functions in `utils.h` and `utils.cpp` for queries that specifically access ProxySQL `global_variables` , and also functions that help with the boilerplate of queries that return a single value:

- `show_variable`
- `show_admin_global_variable`
- `set_admin_global_variable`
- `get_variable_value`

And for handling single value queries there is `mysql_query_ext_val`, which offers boilerplate handling for single values resulset extraction and error handling. Since we have so many, better reuse them, or even reduce them, instead of adding more (like getting rid of the very old ones `show_variable` and `show_admin_global_variable`) or even rewriting some in terms of the newer ones if necessary.

yashwantsahu20 replied:
Fixed it Thanks

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+	time_proxy_compressed = calculate_query_execution_time(proxysql_compression, query);
+	diag("Time taken for query with proxysql with compression level 8: %ld", time_proxy_compressed);
+
+	set_compression_level(proxysql_admin, "3");
```
Individual value check isn't bad, but since the range is small, a simple `for` here can be used to verify the whole range, or at least check the range boundaries, checking the failures outside the range. Also, the table `runtime_global_variables` is never checked, it's possible for a variable to be set, but latter ProxySQL refusing to load it to runtime, since boundaries checks are enforced when `LOAD TO RUNTIME` is issued to `Admin` interface. When verifying if any config table is properly changed, it's important to check the table, but also it's `runtime_*` counterpart.

JavierJF replied:
I see that in the new logic only runtime is checked, it should be fine, but here it was suggested to check both, just wondering, any reason for it? Thanks.

yashwantsahu20 replied:
No specfic reason, added Thanks!

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+		MYSQL_QUERY(proxysql, "INSERT INTO sbtest1 (k, c, pad) SELECT FLOOR(RAND() * 10000), REPEAT('a', 120), REPEAT('b', 60) FROM information_schema.tables LIMIT 1000;");
+	}
+
+	std::string query = "SELECT t1.id id1, t1.k k1, t1.c c1, t1.pad pad1, t2.id id2, t2.k k2, t2.c c2, t2.pad pad2 FROM test.sbtest1 t1 JOIN test.sbtest1 t2 LIMIT 90000000";
```
For performing `SELECT` queries right after `INSERT` queries, we should either:

- Disable potential query rules that can lead to query redirection to replicas.
- Insert query annotation in the `SELECT` themselves that target the `primary` hostgroup.

Otherwise slow replication could lead to errors or not performing the testing as it was intended, since all the required data may not be in the target `replica` server.

yashwantsahu20 replied:
Fixed

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+	MYSQL_ROW row;
+	unsigned long long begin = monotonic_time();
+	unsigned long long row_count = 0;
+	MYSQL_QUERY(mysql, query.c_str());
```
Using this macro here can have unintended consequences. When failing to perform a query, this macro will return `EXIT_FAILURE` (1), so, if for any reason both queries fail to be executed (broken connection for example), the test for checking the execution time:

```
unsigned long diff = abs(time_proxy - time_proxy_compressed);
int performance_diff = (diff * 100) / time_proxy;
```

Would still succeed, as it would be checking `1 - 1 = 0`, and `0` passes the `ok` check right now as a value. Error handling should be encoded here in the return value itself, as they collide with real values from the query/computation.

yashwantsahu20 replied:
Fixed. Now checking for error and returing -1 to indicate error.

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+	unsigned long time_proxy = calculate_query_execution_time(proxysql, query);
+	diag("Time taken for query with proxysql without compression: %ld", time_proxy);
+
+	unsigned long time_proxy_compressed = calculate_query_execution_time(proxysql_compression, query);
```
As a sanity check, would be nice to check that the compressed times are bigger than `0`, specially if `0` is used to encoded failure in `calculate_query_execution_time`, which right now lacks this encoding for failures.

yashwantsahu20 replied:
Now calculate_query_execution_time function will return as -1 incase of error.

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -0,0 +1,184 @@
+#include "command_line.h"
+#include "utils.h"
+
+inline unsigned long long monotonic_time() {
```
I think we have used this function already in several tests files, would be nice to place it in `utils.h`/`utils.cpp`. So we avoid keep copying it around.

yashwantsahu20 replied:
Thanks Fixed

JavierJF reviewed (changes requested):
Comments and changes are focused on the TAP test, there are also some suggestions. Thanks!

yashwantsahu20 committed:
Changes done:
1.  Moved monotonic_time to util
2.  Now calculate_query_execution_time will return -1 incase of error.
3.  Using util function for setting and getting compression variable values.
4.  Added mysql_query_rules

yashwantsahu20 committed:
added more checks

yashwantsahu20 committed:
Added time calculations for MySQL

JavierJF commented on test/tap/tests/mysql-protocol_compression_level-t.cpp:
```diff
@@ -144,10 +152,16 @@ int main(int argc, char** argv) {
 
-	diff = time_proxy_compressed - time_mysql_compressed;
-	performance_diff = (diff * 100) / time_mysql_compressed;
+	time_mysql_without_compressed = calculate_query_execution_time(mysql, query);
```
Hi @yashwantsahu20,

why this `ok` has been turned into a `diag`? Measuring the performance difference between compress/not-compress was one of the motivations for the test right? Also, I consider the test output interesting, this is from CI run #134:

```
[2024-12-13 06:12:54] DEBUG    [proxysql-tester.py:695] msg: Row count: 39690000
[2024-12-13 06:12:54] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:12:54.023576  Time taken for query with proxysql without compression: 111327049
[2024-12-13 06:14:46] DEBUG    [proxysql-tester.py:695] msg: Row count: 39690000
[2024-12-13 06:14:46] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:14:46.266343  Time taken for query with proxysql with compression: 112242706
[2024-12-13 06:14:46] DEBUG    [proxysql-tester.py:695] msg: ok 1 - 2024-12-13 06:14:46.266369 - proxysql with compression performed well compared to without compression. Performance difference: 0 percentage
[2024-12-13 06:15:18] DEBUG    [proxysql-tester.py:695] msg: Row count: 39690000
[2024-12-13 06:15:18] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:15:18.532251  Time taken for query with mysql with compression: 32265856
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: Row count: 39690000
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:15:49.671693  Time taken for query with mysql without compression: 31139396
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:15:49.671718  Time difference for mysql, compression and without compression is: -313005196
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: ok 2 - 2024-12-13 06:15:49.709818 - Run-time default compression level is correct: 3
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: ok 3 - 2024-12-13 06:15:49.720507 - Default compression level is correct: 3
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: ok 4 - 2024-12-13 06:15:49.737985 - Run-time Compression level is set correctly: 8
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: ok 5 - 2024-12-13 06:15:49.747345 - Compression level is set correctly: 8
[2024-12-13 06:17:42] DEBUG    [proxysql-tester.py:695] msg: Row count: 39690000
[2024-12-13 06:17:42] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:17:42.583583  Time taken for query with proxysql with compression level 8: 112836184
[2024-12-13 06:17:42] DEBUG    [proxysql-tester.py:695] msg: ok 6 - 2024-12-13 06:17:42.614726 - Run-time Compression level set correctly: 3
[2024-12-13 06:17:42] DEBUG    [proxysql-tester.py:695] msg: ok 7 - 2024-12-13 06:17:42.617790 - Compression level set correctly: 3
[2024-12-13 06:17:42] DEBUG    [proxysql-tester.py:695] msg: Test took 6 min 40.62 sec
```

The value for the test that was commented doesn't make a lot of sense here:

```
[2024-12-13 06:15:49] DEBUG    [proxysql-tester.py:695] msg: # 2024-12-13 06:15:49.671718  Time difference for mysql, compression and without compression is: -313005196
```

I think an `abs` may be missing in the computation. As a side note, maybe make the values reported for this tests floats, as with integers resolution is lost and the output may look like a failure:

```
[2024-12-13 06:14:46] DEBUG    [proxysql-tester.py:695] msg: ok 1 - 2024-12-13 06:14:46.266369 - proxysql with compression performed well compared to without compression. Performance difference: 0 percentage
```

yashwantsahu20 replied:
Thanksfor your comment.I just thought to remove the test as it was for direct mysql.
Now I have added it back and fixed test value also.

--------------------------------------
ok 1 - 2024-12-19 11:11:19.199704 - proxysql with compression performed well compared to without compression. Performance difference: 0 percentage
Row count: 90000000
# 2024-12-19 11:11:47.122288  Time taken for query with mysql with compression: 27922555
Row count: 90000000
# 2024-12-19 11:12:15.496719  Time taken for query with mysql without compression: 28374382
ok 2 - 2024-12-19 11:12:15.496738 - MySQL with compression performed well compared to without compression. Performance difference: 1 percentage

JavierJF replied:
I see that this wasn't addressed:

>  As a side note, maybe make the values reported for this tests floats, as with integers resolution is lost and the output may look like a failure:
> [2024-12-13 06:14:46] DEBUG    [proxysql-tester.py:695] msg: ok 1 - 2024-12-13 06:14:46.266369 - proxysql with compression performed well compared to without compression. Performance difference: 0 percentage

Using more precision will also allow for other coherence checks (if we are interested) like, increasing the compression level should introduce a penalty. I think this is out-of-scope for the test for now, but having resolution for seeing the actual difference shouldn't hurt.

yashwantsahu20 replied:
Changed to float.

------
# 2025-01-09 11:41:23.519181  Time taken for query with proxysql with compression: 45911204
ok 1 - 2025-01-09 11:41:23.519194 - proxysql with compression performed well compared to without compression. Performance difference: 0.521332 percentage

yashwantsahu20 committed:
Added tests for MySQL with and without compression.

yashwantsahu20 committed:
Merge branch 'v3.0' into v3.0_compression

JavierJF commented on lib/mysql_data_stream.cpp:
```diff
@@ -1234,7 +1234,7 @@ void MySQL_Data_Stream::generate_compressed_packet() {
 			l_free(p2.size,p2.ptr);
 		}
-		int rc=compress(dest, &destLen, source, sourceLen);
+		int rc=compress2(dest, &destLen, source, sourceLen, GloMTH->variables.protocol_compression_level);
```
Looks like I didn't saw this one in the initial review. Since there are per-thread variables introduced in this PR I don't see the need to refer to `GloMTH->variables.protocol_compression_level`. The introduced per-thread variables `mysql_thread___protocol_compression_level` should be used instead. This is a pattern across the code-base with very few exceptions, so we should keep it consistent.

yashwantsahu20 replied:
Thanks fixed.

yashwantsahu20 committed:
1. Using now mysql_thread___protocol_compression_level
2. Changed performance diff to float in tap test

JavierJF commented on lib/mysql_data_stream.cpp:
```diff
@@ -1266,9 +1266,9 @@ void MySQL_Data_Stream::generate_compressed_packet() {
 		destLen2=len2*120/100+12;
 		dest2=(Bytef *)malloc(destLen2+7);
-		rc=compress(dest1+7, &destLen1, (const unsigned char *)p2.ptr, len1);
+		rc=compress2(dest1+7, &destLen1, (const unsigned char *)p2.ptr, len1, GloMTH->variables.protocol_compression_level);
```
Every use of `GloMTH->variables` should be replaced with the per-thread variable. All locations are the same, the previous pin-pointed one was no different from the others.

yashwantsahu20 replied:
Using mysql_thread___protocol_compression_level in all the places

yashwantsahu20 committed:
Using mysql_thread___protocol_compression_level in all the places

JavierJF commented on lib/MySQL_Thread.cpp:
```diff
@@ -2268,6 +2270,7 @@ char ** MySQL_Threads_Handler::get_variables_list() {
 		VariablesPointers_int["client_host_error_counts"]      = make_tuple(&variables.client_host_error_counts,      0,      1024*1024, false);
 		VariablesPointers_int["handle_warnings"]			   = make_tuple(&variables.handle_warnings,				  0,			  1, false);
 		VariablesPointers_int["evaluate_replication_lag_on_servers_load"] = make_tuple(&variables.evaluate_replication_lag_on_servers_load, 0, 1, false);
+		VariablesPointers_int["protocol_compression_level"]    = make_tuple(&variables.protocol_compression_level,    0,      9,         false);
```
Looks like `zlib` allows to setup from `-1` to `9`. `-1` being the default compression `Z_DEFAULT_COMPRESSION`. I think I remember that the change of default was intentional, from `-1` (which right now library claims it's mapped to `6`) to the new default of `3`, please correct me otherwise. But regarding the range, I don't see a reason why we should limit being able to set the library default value back, i.e, `-1`. Maybe this range is something that has been discussed and I don't recall.

yashwantsahu20 replied:
yes ability to set to default can help users. adding it -1 to 9

yashwantsahu20 committed:
changed protocol_compression_level default to  (-1 to 9 )

yashwantsahu20 committed:
Merge branch 'v3.0' into v3.0_compression

renecannao commented:
Can one of the admins verify this patch?
